### PR TITLE
Implement `GetDecisionStatisticsForStage` call

### LINKED ISSUE #182: [Feature]: Implement `GetDecisionStatisticsForStage` call
### Which use case does this feature request relate to?

#73

### Is your request related to a problem? Please describe.

As a project member, I want to see all decision statistics for a specific page.

### Describe the solution you'd like

See [API docs](https://snowballr.informatik.uni-ulm.de/docs/snowballr.html#service-snowballr.SnowballR-GetDecisionStatisticsForStage) for details.

### Describe alternatives you've considered

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by mowi12**
Closes #182 

## What I have made

- Added new `getAllProjectPapersForProject` function + tests
- Added new `getDecisionStatisticsForStage` gRPC call + tests
  - Added validation + tests

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does
not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] (for reviewer) I have checked the implementation against the requirements

### CODE REVIEW

File: src/main/kotlin/se/uulm/snowballr/backend/model/SnowballRException.kt [hunk 1]
```diff
@@ -24,7 +26,7 @@ sealed class SnowballRException(
      * @param entityType The type of the entity that could not be found.
      * @param entityIds The missing entity's ID(s).
      * @param identifierType The type of the entity's identifier. Defaults to [IdentifierType.ID].
-     * @param location The location where the entity could not be found, e.g., the project or paper. Should start with " in ...".
+     * @param location The location where the entity could not be found, e.g., the project or paper. Should start with "in ...".
```
Comment (Merseleo) [inline src/main/kotlin/se/uulm/snowballr/backend/model/SnowballRException.kt]: Nope, this isn't a correct fix. The location must start with a whitespace, as it is directly concatenated and if you remove this space, then you something like "not foundin project 123".

File: src/test/kotlin/se/uulm/snowballr/backend/repository/association/ProjectPaperTableRepoTest.kt [hunk 2]
```diff
@@ -133,6 +133,47 @@ class ProjectPaperTableRepoTest : RepositoryTest(arrayOf(ProjectPaperTable, Proj
             }
     }
 
+    @Nested
+    inner class GetAllProjectPapersForProject {
+        @Test
+        fun `When no project with the given id exists, then an empty list is returned`() = runTest {
```
Comment (Merseleo) [inline src/test/kotlin/se/uulm/snowballr/backend/repository/association/ProjectPaperTableRepoTest.kt]:
```suggestion
        fun `When no project with the given ID exists, then an empty list is returned`() = runTest {
```

File: src/test/kotlin/se/uulm/snowballr/backend/repository/association/ProjectPaperTableRepoTest.kt [hunk 3]
```diff
@@ -133,6 +133,47 @@ class ProjectPaperTableRepoTest : RepositoryTest(arrayOf(ProjectPaperTable, Proj
...
+            val projectId = UUID.randomUUID()
+
+            val result = repo.getAllProjectPapersForProject(projectId)
+
+            assertThat(result).isEmpty()
+        }
+
+        @Test
+        fun `When a project with no project papers exists, then an empty list is returned`() = runTest {
```
Comment (Merseleo) [inline src/test/kotlin/se/uulm/snowballr/backend/repository/association/ProjectPaperTableRepoTest.kt]:
```suggestion
        fun `When the project exists and have no project papers, then an empty list is returned`() = runTest {
```

File: src/test/kotlin/se/uulm/snowballr/backend/repository/association/ProjectPaperTableRepoTest.kt [hunk 4]
```diff
@@ -133,6 +133,47 @@ class ProjectPaperTableRepoTest : RepositoryTest(arrayOf(ProjectPaperTable, Proj
...
+            val paperId = insertPaperAndGetId()
+            insertProjectPaperAndGetId(paperId = paperId, projectId = anotherProject, createdBy = testUserId)
+
+            val result = repo.getAllProjectPapersForProject(projectId)
+
+            assertThat(result).isEmpty()
+        }
+
+        @Test
+        fun `When a project with project papers exists, then the project papers are returned`() = runTest {
```
Comment (Merseleo) [inline src/test/kotlin/se/uulm/snowballr/backend/repository/association/ProjectPaperTableRepoTest.kt]:
```suggestion
        fun `When a project with project papers exists, then all project papers are returned`() = runTest {
```

File: src/test/kotlin/se/uulm/snowballr/backend/validation/ProjectValidatorTest.kt [hunk 5]
```diff
@@ -317,4 +317,47 @@ class ProjectValidatorTest {
...
+        fun `When a valid request is validated, then no issue is returned`() {
+            val request = validGetDecisionStatisticsRequestBuilder.build()
+
+            val result = validateRequest(request)
+
+            EitherAssert.assertThat(result).isRight()
+        }
+
+        @Test
+        fun `When the stage is negative, then the 'OutOfRangeValue' issue is returned`() {
```
Comment (Merseleo) [inline src/test/kotlin/se/uulm/snowballr/backend/validation/ProjectValidatorTest.kt:461]: And what happens, if the stage value is way to big?

File: src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    @Test
+    fun `When a server admin retrieves the decision statistics, then no exception is thrown`() = runTest {
+        val user = DataBuilder.createExampleUser(role = UserOuterClass.UserRole.USER_ROLE_ADMIN)
+        val project = DataBuilder.createExampleProject()
+
+        val request = ProjectOuterClass.Project.Information.DecisionStatistics.Get
+            .newBuilder()
+            .setProjectId(project.id.toString())
+            .setStage(0)
+            .build()
```
Comment (Merseleo) [inline src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt]: You the same or a very similar request a lot of times, isn't it possible to extract it to a requestBuilder function or something similar?

File: src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt [hunk 7]
```diff
@@ -194,4 +207,76 @@ class ProjectService(
...
+        request: GrpcProjectDecisionStatistics.Get,
+    ): GrpcProjectDecisionStatistics = withUser(userRepo) { currentUser ->
+        val projectId = parseUUID(request.projectId, EntityType.PROJECT)
+
+        isAllowedToReadProject(projectMemberRepo).checkFor(currentUser, projectId)
+
+        val project = repo.getProjectById(projectId).getOrThrow()
+        val currentStage = project.currentStage
+
+        if (request.stage > currentStage) {
```
Comment (Merseleo) [inline src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt]: I think you should test, whether the requested stage is greater than the max stage.

File: src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt [hunk 8]
```diff
@@ -194,4 +207,76 @@ class ProjectService(
...
+    private suspend fun createStatistics(projectId: UUID, stage: Long): List<GrpcProjectDecisionStatistics.Statistic> {
+        var acceptedCount = 0L
+        var declinedCount = 0L
+        var unreviewedCount = 0L
+        var inReviewCount = 0L
+
+        projectPaperRepo.getAllProjectPapersForProject(projectId).forEach { projectPaper ->
+            if (projectPaper.stage == stage) {
+                when (projectPaper.decision) {
+                    ProjectOuterClass.PaperDecision.PAPER_DECISION_ACCEPTED -> acceptedCount++
```
Comment (Merseleo) [inline src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt]:
```suggestion
                    PaperDecision.PAPER_DECISION_ACCEPTED -> acceptedCount++
```
If you import the `PaperDecision` directly, would reduce the line length a bit.

File: src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt [hunk 9]
```diff
@@ -194,4 +207,76 @@ class ProjectService(
...
+
+        val statistics = createStatistics(projectId, request.stage)
+        GrpcProjectDecisionStatistics
+            .newBuilder()
+            .addAllStatistics(statistics)
+            .build()
+    }
+
+    /**
+     * Creates a list of [GrpcProjectDecisionStatistics.Statistic] objects for the given [projectId] and [stage].
```
Comment (Merseleo) [inline src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt]:
```suggestion
     * Creates a list of [GrpcProjectDecisionStatistics.Statistic] objects for the given [stage] in the project with the ID [projectId].
```

File: src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt [hunk 10]
```diff
@@ -194,4 +207,76 @@ class ProjectService(
...
+     * @return A list of [GrpcProjectDecisionStatistics.Statistic] objects.
+     */
+    private suspend fun createStatistics(projectId: UUID, stage: Long): List<GrpcProjectDecisionStatistics.Statistic> {
+        var acceptedCount = 0L
+        var declinedCount = 0L
+        var unreviewedCount = 0L
+        var inReviewCount = 0L
+
+        projectPaperRepo.getAllProjectPapersForProject(projectId).forEach { projectPaper ->
+            if (projectPaper.stage == stage) {
```
Comment (Merseleo) [inline src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt]: One general idea (not necessary): Couldn't you add the stage as parameter for the repo call and then add a second where statement to the SQL statement and only request the papers for this stage?

File: src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt
Comment (Merseleo) [inline src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt:1]: I asked our friend and got a suggestion:
```kotlin
private suspend fun createStatistics(
    projectId: UUID,
    stage: Long,
): List<GrpcProjectDecisionStatistics.Statistic> {
    val counts = projectPaperRepo.getAllProjectPapersForProject(projectId)
        .asSequence()
        .filter { it.stage == stage }
        .groupingBy { it.decision }
        .eachCount()
        .mapValues { it.value.toLong() }

    fun stat(decision: ProjectOuterClass.PaperDecision) =
        GrpcProjectDecisionStatistics.Statistic.newBuilder()
            .setDecision(decision)
            .setCount(counts[decision] ?: 0)
            .build()

    return listOf(
        ProjectOuterClass.PaperDecision.PAPER_DECISION_ACCEPTED,
        ProjectOuterClass.PaperDecision.PAPER_DECISION_DECLINED,
        ProjectOuterClass.PaperDecision.PAPER_DECISION_UNREVIEWED,
        ProjectOuterClass.PaperDecision.PAPER_DECISION_IN_REVIEW,
    ).map(::stat)
}
```
You do not have to adapt it, but I think there are some nice ideas that could be used. Use what you consider as useful.

File: src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        mockCurrentUser(user)
+        coEvery { projectMemberRepoMock.getProjectMembers(project.id) } returns listOf(projectMember)
+        coEvery { projectRepoMock.getProjectById(project.id) } returns Result.success(project)
+        coEvery { projectPaperRepoMock.getAllProjectPapersForProject(project.id) } returns emptyList()
+
+        assertDoesNotThrow { mainService.getDecisionStatisticsForStage(request) }
+    }
+
+    @Test
+    fun `When a non-member retrieves the decision statistics, then an UnauthorizedException is thrown`() = runTest {
```
Comment (Merseleo) [inline src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt]:
```suggestion
    fun `When a non member retrieves the decision statistics, then an UnauthorizedException is thrown`() = runTest {
```
We always used it this way.

File: src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+        mockCurrentUser(user)
+        coEvery { projectMemberRepoMock.getProjectMembers(any()) } returns emptyList()
+        coEvery { projectRepoMock.getProjectById(any()) } returns Result.failure(TestSpecificException())
+
+        assertThrows<TestSpecificException> { mainService.getDecisionStatisticsForStage(request) }
+    }
+
+    @Test
+    fun `When an invalid stage is requested, then a StageNotFoundException is thrown`() = runTest {
```
Comment (Merseleo) [inline src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt]: In one case we state that TestSpecificException is thrown and in one case the concrete StageNotFoundException? Is that correct and I dont understand the system?

Comment (Merseleo) [inline src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt]: Maybe it would be better to set an explicit stage in the project setup and then test it for stage + 1 so test the "edge" case.

File: src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+            .setProjectId(project.id.toString())
+            .setStage(stage)
+            .build()
+
+        mockCurrentUser(user)
+        coEvery { projectMemberRepoMock.getProjectMembers(project.id) } returns emptyList()
+        coEvery { projectRepoMock.getProjectById(project.id) } returns Result.success(project)
+        coEvery { projectPaperRepoMock.getAllProjectPapersForProject(project.id) } returns projectPapers
+
+        val statistics = mainService.getDecisionStatisticsForStage(request)
```
Comment (Merseleo) [inline src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt]: Maybe wrap it again with `assertDoesNotThrow`

**Review by Merseleo — CHANGES_REQUESTED** [review]
Good work, I think there is not that much to adapt. In general I manually tested the call and everything worked correctly!

File: src/main/kotlin/se/uulm/snowballr/backend/model/SnowballRException.kt (hunk 1 above)
Comment (mowi12) [reply src/main/kotlin/se/uulm/snowballr/backend/model/SnowballRException.kt]: Oh well. Dropped this part.

File: src/test/kotlin/se/uulm/snowballr/backend/repository/association/ProjectPaperTableRepoTest.kt (hunk 3 above)
Comment (mowi12) [reply src/test/kotlin/se/uulm/snowballr/backend/repository/association/ProjectPaperTableRepoTest.kt]: You're suggestions wasn't quite correct. I fixed it.

File: src/test/kotlin/se/uulm/snowballr/backend/validation/ProjectValidatorTest.kt (hunk 5 above)
Comment (mowi12) [reply src/test/kotlin/se/uulm/snowballr/backend/validation/ProjectValidatorTest.kt:461]: Nothing.
If the value is not from the type Long, then the client shouldn't be able to send the message.
And if the value is in range of the stages available will not be checked here but in the service layer, as it is not possible here.
The only thing the validator can check, is if the value is negative, since there are no negative stages.

File: src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt (hunk 6 above)
Comment (mowi12) [reply src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt]: I added a default valid request. But in my opinion this doesn't change much, since I always need to specify the `projectId` and in two cases it literally doesn't change anything as both values need to be updated.

I still included it, since there are 3-4 lines removed through the change.

File: src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt (hunk 7 above)
Comment (mowi12) [reply src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt]: You're right. The current stage could be not the last one.

File: src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt (hunk 9 above)
Comment (mowi12) [reply src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt]: Yes, why not.

File: src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt (hunk 10 above)
Comment (mowi12) [reply src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt]: I also thought about that, but if I thought correctly about that, this will not result in a more efficient database call, since all projectPapers will be fetched, but only the ones that are that are in the correct stage will be returned.

This way, we can use this call for any kind of action, where all project papers are needed.

But I will refactor the check, since this can be done with a `filter()`.

File: src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt
Comment (mowi12) [reply src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt:1]: Will do!

File: src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt (hunk 11 above)
Comment (mowi12) [reply src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt]: In my opinion, both are correct, but since we used your variant more often in the project, I will include this for the sake of consistency.

File: src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt (hunk 12 above)
Comment (mowi12) [reply src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt]: Yes, that's correct.

The `TestSpecificException` is always returns if we would return a `NotFoundException`, which we would get in the first case. We do this, because the `NotFoundException` expects a lot of parameters in the constructor, which are all not relevant for the test.

In the second case, we do not use the generic `NotFoundException`, but an exception which we do not need to mock, so we do expect the actual exception.

File: src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt (hunk 13 above)
Comment (mowi12) [reply src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt]: Just forgot that.

File: src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt (hunk 12 above)
Comment (mowi12) [reply src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt]: I'm not quite sure which edge case you mean.

**Comment by mowi12** [comment]
I had some problems while rebasing. Because of that I created the commits from scratch, and lost all of the new review commits. I hope this isn't too bad, sorry!

File: src/test/kotlin/se/uulm/snowballr/backend/repository/association/ProjectPaperTableRepoTest.kt (hunk 3 above)
Comment (Merseleo) [reply src/test/kotlin/se/uulm/snowballr/backend/repository/association/ProjectPaperTableRepoTest.kt]: Great.

File: src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt (hunk 11 above)
Comment (Merseleo) [reply src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt]: Ah ok, then I will not highlight this anymore if both are correct.

File: src/test/kotlin/se/uulm/snowballr/backend/validation/ProjectValidatorTest.kt (hunk 5 above)
Comment (Merseleo) [reply src/test/kotlin/se/uulm/snowballr/backend/validation/ProjectValidatorTest.kt:461]: Ah ok, i understand.

File: src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt (hunk 7 above)
Comment (Merseleo) [reply src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt]: But the entire stage system is so weird in my opinion ...

File: src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt (hunk 10 above)
Comment (Merseleo) [reply src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt]: Kind of correct, however, databases can filter the data more efficiently, at least this would be my guess, because they are optimized for this task. However, we aren't considering performance in the entire application anyway, so lets ignore this.

File: src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt (hunk 6 above)
Comment (Merseleo) [reply src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt]: If you do not create the user and project always in a setup method and doesn't know the project id, you are right, then it doesn't make really sense. I would extract more than that, but probably we should not discuss this here. When the missing calls are implemented and we test the application we can do such improvements if necessary.

File: src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt (hunk 12 above)
Comment (Merseleo) [reply src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt]: Ok.

Comment (Merseleo) [reply src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt]: The project you create only has the stage 0 as current and max stage. But you test it with a stage 42. If you want to test such things (at least this is the way I learned it in "Software Security Testing") like the access using an index number, then you actually test the border value (if inclusive no failure, if exclusive failure) and the two values around the border. 
So in your case: Assume the project has current stage 0, max stage 1, then for the stage 0 and 1, no exception should be thrown, but for 2. In case somebody changes the condition from > maxStage to >= max stage you would recognize this behavior in your tests, with my suggestion you would detect it, because stage 1 shouldn't throw an error.

File: src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+        val statistics = assertDoesNotThrow { mainService.getDecisionStatisticsForStage(request) }
+
+        val statsByDecision = statistics.statisticsList.associateBy { it.decision }
+        paperCountsByDecision.forEach { (decision, expectedCount) ->
+            val actual = statsByDecision[decision]?.count ?: -1
+            assertEquals(
+                expectedCount,
+                actual.toInt(),
+            )
+        }
```
Comment (Merseleo) [inline src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt:181]:
```suggestion
        val paperCountsByDecision = mapOf(
            PaperDecision.PAPER_DECISION_ACCEPTED to 1L,
            PaperDecision.PAPER_DECISION_DECLINED to 2L,
            PaperDecision.PAPER_DECISION_IN_REVIEW to 3L,
            PaperDecision.PAPER_DECISION_UNREVIEWED to 4L,
        )

        val projectPapers = paperCountsByDecision.flatMap { (decision, count) ->
            (1..count).map {
                DataBuilder.createExampleProjectPaper(
                    projectId = project.id,
                    stage = stage,
                    decision = decision,
                )
            }
        }

        val request = validRequestBuilder
            .setProjectId(project.id.toString())
            .setStage(stage)
            .build()

        mockCurrentUser(user)
        coEvery { projectMemberRepoMock.getProjectMembers(project.id) } returns emptyList()
        coEvery { projectRepoMock.getProjectById(project.id) } returns Result.success(project)
        coEvery { projectPaperRepoMock.getAllProjectPapersForProject(project.id) } returns projectPapers

        val statistics = assertDoesNotThrow { mainService.getDecisionStatisticsForStage(request) }

        val statsByDecision = statistics.statisticsList.associateBy { it.decision }
        paperCountsByDecision.forEach { (decision, expectedCount) ->
            val actual = statsByDecision[decision]?.count ?: -1L
            assertEquals(expectedCount, actual)
        }
```

Comment (Merseleo) [reply src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt:181]: Optional. If you dont think it is an improvement, close it.

Comment (Merseleo) [reply src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt:181]: Furthermore: if you want to make the tests more robust, add some papers in another stage, because if somebody changes the stage filter, then the actual count would be bigger and it would be noticed.

**Review by Merseleo — CHANGES_REQUESTED** [review]
> I had some problems while rebasing. Because of that I created the commits from scratch, and lost all of the new review commits. I hope this isn't too bad, sorry!

No problem!

File: src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt (hunk 6 above)
Comment (mowi12) [reply src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt]: Ok. Then I keep it as it is now.

File: src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt (hunk 12 above)
Comment (mowi12) [reply src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt]: Thank you for the explanation. Now this makes a lot more sense!
I added the missing test cases.

File: src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt (hunk 14 above)
Comment (mowi12) [reply src/test/kotlin/se/uulm/snowballr/backend/service/project/GetDecisionStatisticsForStageTest.kt:181]: Good idea. I've added two additional papers, which should not be included in the final count.

**Review by Merseleo — APPROVED** [review]
Nothing to complain about, lets merge.
